src/dune_rpc_server/dune_rpc_server.ml

```diff
@@ -246,27 +314,77 @@ module H = struct
     type 's r_handler =
       | R : ('s, 'a, 'b) callback * ('a, 'b) Decl.request -> 's r_handler
 
+    type 's sub =
+      | Sub :
+          'a Sub.t
+          * ('s Session.t -> 'init Fiber.t)
+          * ('s Session.t -> 'init -> 'a Subscription.t -> unit Fiber.t)
+          -> 's sub
+
     type 's t =
       { mutable notification_handlers : 's n_handler list
       ; mutable request_handlers : 's r_handler list
+      ; mutable subscriptions : 's sub String.Map.t
       ; on_init : 's Session.t -> Initialize.Request.t -> 's Fiber.t
       ; on_terminate : 's Session.t -> unit Fiber.t
       ; version : int * int
       }
 
+    let on_subscribe_request subscriptions =
+      let callback =
+        let info = public ~since:(3, 0) () in
+        let f session req_id what =
+          let req_id = Option.value_exn req_id in
+          match String.Map.find subscriptions what with
+          | None ->
+            let err =
+              let payload = Sexp.record [ ("what", Sexp.Atom what) ] in
+              Response.Error.create ~kind:Invalid_request ~payload
+                ~message:"Unknown subscription" ()
+            in
+            raise (Response.Error.E err)
+          | Some (Sub (sub, on_subscribe, init_subscription)) ->
+            let* init = on_subscribe session in
+            let subscription = Session.create_subscription session req_id sub in
+            (* We need to send a response and proceed here. That's not possible
```

Comment:
I don't understand this comment. "We need to send a response" to achieve what? What's wrong with the current behavior?

Reply:
To implement an RPC handler for a request, one must provide a callback of the type (omitting the gadt noise):

```ocaml
request -> response Fiber.t
```

But this callback limits us from scheduling work after the response has been successfully sent. For example, we cannot do:

```ocaml
let* () = (* response has been sent to the client *) in
do_something ()
```

I've worked around this but rewriting the above as:

```ocaml
let* () = Fiber.Pool.task session.pool ~f:(fun () -> do_something ()) in
response
```

But the semantics are not quite the same. `do_something` is scheduled if the client doesn't receive the response.

My comment alluded to changing the callback interface to be:

```ocaml
request -> (response -> unit Fiber.t) -> unit Fiber.t
```

Where the continuation parameter must be called once to send the response. The above could then be written as:

```ocaml
let*() = k response in
do_something ()
````

With the desired semantics.

Reply:
Ok. Sounds like the problem is the race between sending the initial response and sending the updates.
Can you please clarify the comment? Also, it's still not clear to me if this is a "code smell" kind of thing, or a real bug.

Reply:
By the way, I was slightly surprised when I saw that subscriptions are implemented on top of requests and notifications instead of being their own kind of rpc (how it's done in Async_rpc). I guess that's a good thing if it can be done cleanly, but this issue sounds like it's arising from that encoding.

Reply:
> I guess that's a good thing if it can be done cleanly, but this issue sounds like it's arising from that encoding.

That's true, but I pointed out a real problem with how we allow to implement request/responses. It's natural to want to do things after the response has been sent to the client. I think this deficiency in the api should be solved. The CPS based solution I described above is used in a few libraries I'm familiar with to solve this problem btw (Haskell's WAI for example).